clang/docs/SYCLFERFCIntegrationHeaderFooter.md

```diff
@@ -0,0 +1,179 @@
+        // Call smplr sampler's init function
+        sycl::accessor::init(&local.smplr, smpData);
+
+        // Call the kernel body
+        callee(&local);
+    }
+
+    spir_func void callee(struct FuncObj* this)
+    {
+        // body of the kernel invocation
+    }
+```
+
+For details of this transformation, see 'Lowering of SYCL-Kernel'. // TODO: Add link
+
+The integration header then describes the fields of the kernel objects as follows:
+
+```
+namespace sycl {
+  inline namespace _V1 {
+    namespace detail {
+      // names of all kernels defined in the corresponding source
+      static constexpr
+      const char* const kernel_names[] = {
+        "_ZTSZ4mainE12first_kernel"  // mangled name associated with first_kernel
+                                     // demangles to 'typeinfo name for main::first_kernel'
+
+      };
+      // array representing signatures of all kernels defined in the
+      // corresponding source
+      static constexpr
+      const kernel_param_desc_t kernel_signatures[] = {
+        //--- _ZTSZ4mainE12first_kernel
+        { kernel_param_kind_t::kind_std_layout, 4, 0 },
+        { kernel_param_kind_t::kind_std_layout, 8, 4 },
+        { kernel_param_kind_t::kind_accessor, 4062, 12 },
+        { kernel_param_kind_t::kind_sampler, 8, 24 },
+
+        { kernel_param_kind_t::kind_invalid, -987654321, -987654321 },
+      };
+    } // namespace detail
+  } // namespace _V1
+} // namespace sycl
+```
+
+For each kernel, a mangled name associated with it, is stored in the kernel_names array.
+
+And for each kernel, each of its parameter is described in the kernel_signatures array.   The description consists of the following parts:
+
+* an encoding of the type that is captured
+* its size and its offset
+
+In the above example, the first two parameters are standard layout types (the int 'i' and the struct 'test_s'), the latter two though standard layout types are types which need special handling - accessor and sampler respectively.   The last entry acts as a terminating entry.
+
+This allows the run-time library to recreate the function object on the device.
+
+The full set of encoding is:
+
+* kind_accessor  for sycl accessor types
+* kind_std_layout for standard layout types
+* kind_sampler for sycl sampler types
+* kind_pointer for pointer types
+* kind_specialization_constants_buffer for initializing specialization constants
+* kind_stream for sycl stream types and
+* kind_invalid used to indicate the terminal value in the signatures array.
+
+
+# Integration Footer
+
+
+We propose an Integration Footer primarily to support SYCL specialization constants.
+
+The SYCL 2020 Specification defines a specialization constant as a constant variable where the value is not known until compilation of the SYCL kernel function.  In order to accommodate both implementations that have native support for this as well as those that do not, we propose the following mechanism:
+
```

Comment:
The integration header description above nicely leads with sample code and then shows what ends up in the integration header. I like that. Can we do the same here? E.g., move the commented out `Wrapper` class definition below to here?

Reply:
Done.   Thanks for the suggestion.